### PR TITLE
[MI-2935]:Fixed issue #378 'Check for webhook while subscribing'

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
https://github.com/mattermost/mattermost-plugin-github/issues/378

### CODE REVIEW

File: server/plugin/command.go [hunk 1]
```diff
@@ -6,23 +6,26 @@ import (
...
+	featureIssues                 = "issues"
+	featurePulls                  = "pulls"
+	featurePullsMerged            = "pulls_merged"
+	featurePushes                 = "pushes"
+	featureCreates                = "creates"
+	featureDeletes                = "deletes"
+	featureIssueComments          = "issue_comments"
+	featurePullReviews            = "pull_reviews"
+	featureStars                  = "stars"
+	NoWebhookFoundMsg             = "\nNo webhook was found for this repository or organization. To create one, enter the following slash command `/github setup webhook`"
```
Comment (ayusht2810) [inline server/plugin/command.go]: i think we should define a different constants for error messages

Comment (ayusht2810) [inline server/plugin/command.go]:
```suggestion
	NoWebhookFoundMsg             = "\nNo webhook was found for this repository or organization. To create one, run the following slash command `/github setup webhook`"
```

Comment (ayusht2810) [reply server/plugin/command.go]: or `execute` maybe?

File: server/plugin/command.go [hunk 2]
```diff
@@ -265,6 +268,48 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+		var githubResponse *github.Response
+		var err error
+
+		if repo == "" {
+			githubHooks, githubResponse, err = githubClient.Organizations.ListHooks(ctx, owner, opt)
+		} else {
+			githubHooks, githubResponse, err = githubClient.Repositories.ListHooks(ctx, owner, repo, opt)
+		}
+		if err != nil {
+			p.API.LogWarn("Not able to get the list of webhooks", "Owner", owner, "Repo", repo, "Error", err)
```
Comment (ayusht2810) [inline server/plugin/command.go]:
```suggestion
			p.API.LogWarn("Not able to get the list of webhooks", "Owner", owner, "Repo", repo, "Error", err.Error())
```

File: server/plugin/command.go [hunk 3]
```diff
@@ -265,6 +268,48 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+			// Breaking from the loop if the repo or org is not found
+			if strings.Contains(err.Error(), "404 Not Found") {
+				isWebhook = true
+				break
+			} else {
+				return isWebhook, err
+			}
+		}
+		for _, hook := range githubHooks {
```
Comment (ayusht2810) [inline server/plugin/command.go]:
```suggestion
		}
		
		for _, hook := range githubHooks {
```

File: server/plugin/command.go [hunk 4]
```diff
@@ -265,6 +268,48 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+			if strings.Contains(hook.Config["url"].(string), p.getSiteURL()) {
+				isWebhook = true
+			}
+		}
+		if githubResponse.NextPage == 0 {
```
Comment (ayusht2810) [inline server/plugin/command.go]:
```suggestion
		}
		
		if githubResponse.NextPage == 0 {
```

File: server/plugin/command.go (hunk 1 above)
Comment (Kshitij-Katiyar) [reply server/plugin/command.go]: The msg is given by mike https://mm.brightscout.com/freelancers/pl/wgjacqcoz3nu8n4mdgndedwhte

**Review by ayusht2810 — APPROVED** [review]

File: server/plugin/command.go [hunk 5]
```diff
@@ -265,6 +271,49 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+		var githubHooks []*github.Hook
+		var githubResponse *github.Response
+		var err error
+
+		if repo == "" {
+			githubHooks, githubResponse, err = githubClient.Organizations.ListHooks(ctx, owner, opt)
+		} else {
+			githubHooks, githubResponse, err = githubClient.Repositories.ListHooks(ctx, owner, repo, opt)
+		}
+		if err != nil {
```
Comment (raghavaggarwal2308) [inline server/plugin/command.go:291]:
```suggestion
		if err != nil {
```
```suggestion

		if err != nil {
```

File: server/plugin/command.go [hunk 6]
```diff
@@ -25,6 +26,11 @@ const (
 	featureStars         = "stars"
 )
 
+const (
+	ErrorNoWebhookFoundMsg        = "\nNo webhook was found for this repository or organization. To create one, enter the following slash command `/github setup webhook`"
```
Comment (raghavaggarwal2308) [inline server/plugin/command.go]:
```suggestion
	ErrorNoWebhookFound        = "\nNo webhook was found for this repository or organization. To create one, enter the following slash command `/github setup webhook`"
```

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/command.go [hunk 7]
```diff
@@ -265,6 +271,49 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
 	return txt
 }
 
+func (p *Plugin) getWebhookListForRepoOrOrg(githubClient *github.Client, repo, owner string, ctx context.Context) (bool, error) {
+	isWebhook := false
+
```
Comment (manojmalik20) [inline server/plugin/command.go]:
```suggestion
```

File: server/plugin/command.go [hunk 8]
```diff
@@ -265,6 +271,49 @@ func (p *Plugin) handleSubscriptionsList(_ *plugin.Context, args *model.CommandA
...
+				isWebhook = true
+				break
+			} else {
+				return isWebhook, err
+			}
+		}
+
+		for _, hook := range githubHooks {
+			if strings.Contains(hook.Config["url"].(string), p.getSiteURL()) {
+				isWebhook = true
```
Comment (manojmalik20) [inline server/plugin/command.go:304]: Shouldn't we break after this line?

File: server/plugin/command.go [hunk 9]
```diff
@@ -325,7 +374,17 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
...
-		return fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+		var subOrgMsg = fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+
+		isWebhook, err := p.getWebhookListForRepoOrOrg(githubClient, repo, owner, ctx)
+		if err != nil {
+			return errors.Wrap(err, "failed to get the list of webhooks").Error()
+		}
+
+		if !isWebhook {
+			subOrgMsg += ErrorNoWebhookFoundMsg
```
Comment (manojmalik20) [inline server/plugin/command.go]:
```suggestion
			subOrgMsg = ErrorNoWebhookFoundMsg
```

File: server/plugin/command.go [hunk 10]
```diff
@@ -325,7 +374,17 @@ func (p *Plugin) handleSubscribesAdd(_ *plugin.Context, args *model.CommandArgs,
 			return err.Error()
 		}
 
-		return fmt.Sprintf("Successfully subscribed to organization %s.", owner)
+		var subOrgMsg = fmt.Sprintf("Successfully subscribed to organization %s.", owner)
```
Comment (manojmalik20) [inline server/plugin/command.go]:
```suggestion
		subOrgMsg := fmt.Sprintf("Successfully subscribed to organization %s.", owner)
```

File: server/plugin/command.go (hunk 9 above)
Comment (Kshitij-Katiyar) [reply server/plugin/command.go]: Here we are appending the previous message with the new one.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**